### ISSUE TITLE
Emulator does not accept more than 1 mb of ram for  gravis ultrasound interwave

### ISSUE THREAD

**Issue by kotenok2000** [joncampbell123/dosbox-x#3336]
### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

### Have you checked that no other similar bug report(s) already exists?

- [X] I have searched and didn't find any similar issues.

### What operating system(s) this bug have occurred on?

Microsoft Windows [Version 10.0.22000.556]

### What version(s) of DOSBox-X have this bug?

0.83.23 sdl1

### Describe the bug

When i try to set Gravis Ultrasound Plug & Play memory to 16384 kb configuration tool does not accept it and changes it to 1024 kb.

### Expected behavior

Wikipedia says that gravis ultrasound pnp ram should support up to 16 mb of ram, https://en.wikipedia.org/wiki/Gravis_Ultrasound#:~:text=In%202014%20a%20RAM%20adapter%20for%2072%2Dpin%20SIMM%20was%20produced%20by%20retro%2Dcomputer%20enthusiasts%20that%20made%20it%20possible%20to%20install%2016%20MB%20of%20RAM%20on%20the%20%27Pro%27%20version%20without%20any%20modifications%20to%20the%20card.

### Steps to reproduce the behaviour

1. Go to configuration tool.
2. Press GUS
3. Set gustype interwave
4. Set memory size 16384
5. press ok
6. press gus again
7. see that memory was changed to 1024

**Comment by BridgeHeadland** [comment]
DOSBox-X at this time only emulates up to the Gravis Ultrasound MAX. 😕

**Comment by kotenok2000** [comment]
So interwave is just an alias for max for now?

**Comment by rderooy** [comment]
Interwave emulation currently does not work indeed. It is just there for developers to work on, but it has been dormant for a long time.

86box also only supports the classic and the max.

**Comment by rderooy** [comment]
Forgot to mention that the MAX emulation is also not complete. It is missing emulation of the WSS codec chip.

**Comment by BridgeHeadland** [comment]
> Glemte å nevne at MAX-emuleringen heller ikke er komplett. Den mangler emulering av WSS-kodekbrikken.

Will the Interwave emulation be finished or done before Christmas or something? I miss playing Doom PWADs in MBF with GUSPNP. As I said many times, I also miss an unstretched openglnb, such as openglnb in SVN Daum (when aspect is set to false), will something be done with it, like before Christmas? I hope I do not bother.

**Comment by BridgeHeadland** [comment]
Has there been more done with Max and Interwave since last (March)? If there are any specific files you need for the virtual Z drive, for Max and Interwave to work, I can do some research on the web to find some, to save time. Or maybe I should just download driver files and stuff to my virtual C drive? Sometimes I read through things that have something to do with DOS, but unfortunately I won't understand more than the little I already understand, and that's where it stops.

**Comment by rderooy** [comment]
Nobody is actively working on enhancing GUS emulation to my knowledge.

You can check for changes to the GUS emulation here:
https://github.com/joncampbell123/dosbox-x/commits/master/src/hardware/gus.cpp

**Comment by BridgeHeadland** [comment]
@rderooy It is a long time.

**Comment by JayQ2K** [comment]
Checking lines 2533 and 2534 I read something that might need some extra light.

I read "FIXME What do these numbers mean?"...well (from the (Gravis Ultrasound Installation Guide"):
_SET ULTRA16=32c,0,0,1,0 (for UltraSound MAX only)
Tells applications what settings to use for 16-bit recording. 
The parameters are: 16-bit base port address, not used, not used, MAX identifier, not used._

Besides that, seemingly for the INTERWAVE emulation, the set ULTRA16 option is also created, while this should be the set INTERWAVE option: "_...automatically detect that the UltraSound PnP or Max is installed. It does this by looking at the SET INTERWAVE (PnP) or SET ULTRA16 (Max) environment variables respectively._"

Concerning MAX emu not complete: "_CS4231 provides support for Windows Sound System specs, although the IO port range doesn't match the WSS hardware_"...so in real life this don't work on a MAX anyway. You would actually need the classic version with the separate daugther board to get that all to work.
So if it's only that what makes MAX emu incomplete, you could also say it works as designed.

**Comment by JayQ2K** [comment]
More broadly I guess it would make sense to split gus.cpp into something like gus.cpp and guspnp.cpp.
GF1 based cards do not need PnP so it cleans up nicely to get these distinct.
The interwave setting in dosbox-x.conf would then just use guspnp.cpp.

That way you can then clean up most gus.cpp code to at some point emulate a GF1 that actually has proper WSS settings to bring it all together.

**Comment by joncampbell123** [comment]
Just to point something out, Gravis Ultrasound MAX cards also have a WSS chip. If WSS emulation is to be implemented here, it should be in a more general fashion between the Interwave and "classic" emulation. The real GUS MAX I use for testing has a WSS chip and that is what, supposedly, the ULTRA16 environment variable is all about.

There are intricate details in GF1 emulation based on real hardware as well that are reportedly not fully emulated by the Interwave cards. As I've stated elsewhere, the real GF1 chipset seems to behave like a rather dumb loop in hardware, iterating per voice, but equipped with the high performance sample fetch, interpolate, and voice volume calculations needed to do it's job. Reportedly the Interwave mostly faithfully emulates the GF1 behavior but with differences that might break some games. They are different enough that, yes, perhaps there should be a guspnp.cpp.

**Comment by JayQ2K** [comment]
> Just to point something out, Gravis Ultrasound MAX cards also have a WSS chip. If WSS emulation is to be implemented here, it should be in a more general fashion between the Interwave and "classic" emulation. The real GUS MAX I use for testing has a WSS chip and that is what, supposedly, the ULTRA16 environment variable is all about.
> 
Correct. The ULTRA16 variable is for your system to know that a WSS Compatible Codec is installed. For GUS you should therefore get (or set) that variable when you run a CS4231 chip (daughterboard or MAX) or CS4231 compatible (Interwave).
